File: src/test/java/it/gabrieletondi/telldontaskkata/useCase/OrderApprovalUseCaseTest.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.doubles.TestOrderRepository;
import org.junit.Test;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;

public class OrderApprovalUseCaseTest {
    private final TestOrderRepository orderRepository = new TestOrderRepository();
    private final OrderApprovalUseCase useCase = new OrderApprovalUseCase(orderRepository);

    @Test
    public void approvedExistingOrder() {
        Order initialOrder = Order.of(1, OrderStatus.CREATED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, true);
        useCase.run(request);

        final Order savedOrder = orderRepository.getSavedOrder();
        assertThat(savedOrder.getStatus(), is(OrderStatus.APPROVED));
    }

    @Test
    public void rejectedExistingOrder() {
        Order initialOrder = Order.of(1, OrderStatus.CREATED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, false);
        useCase.run(request);

        final Order savedOrder = orderRepository.getSavedOrder();
        assertThat(savedOrder.getStatus(), is(OrderStatus.REJECTED));
    }

    @Test(expected = RejectedOrderCannotBeApprovedException.class)
    public void cannotApproveRejectedOrder() {
        Order initialOrder = Order.of(1, OrderStatus.REJECTED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, true);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
    }

    @Test(expected = ApprovedOrderCannotBeRejectedException.class)
    public void cannotRejectApprovedOrder() {
        Order initialOrder = Order.of(1, OrderStatus.APPROVED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, false);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
    }

    @Test(expected = ShippedOrdersCannotBeChangedException.class)
    public void shippedOrdersCannotBeApproved() throws Exception {
        Order initialOrder = Order.of(1, OrderStatus.SHIPPED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, true);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
    }

    @Test(expected = ShippedOrdersCannotBeChangedException.class)
    public void shippedOrdersCannotBeRejected() throws Exception {
        Order initialOrder = Order.of(1, OrderStatus.SHIPPED);
        orderRepository.addOrder(initialOrder);

        OrderApprovalRequest request = OrderApprovalRequest.of(1, false);
        useCase.run(request);

        assertThat(orderRepository.getSavedOrder(), is(nullValue()));
    }
}

File: typescript/test/useCase/OrderApprovalUseCase.test.ts
import Order from "../../src/domain/Order";
import OrderStatus from "../../src/domain/OrderStatus"
import OrderApprovalRequest from "../../src/useCase/OrderApprovalRequest";
import TestOrderRepository from "../doubles/TestOrderRepository";
import OrderApprovalUseCase from "../../src/useCase/OrderApprovalUseCase";
import RejectedOrderCannotBeApprovedError from "../../src/useCase/RejectedOrderCannotBeApprovedError";
import ApprovedOrderCannotBeRejectedError from "../../src/useCase/ApprovedOrderCannotBeRejectedError";
import ShippedOrdersCannotBeChangedError from "../../src/useCase/ShippedOrdersCannotBeChangedError";

describe('OrderApprovalUseCase should', () => {
    let orderRepository: TestOrderRepository;
    let useCase: OrderApprovalUseCase;

    beforeEach(() => {
        orderRepository = new TestOrderRepository();
        useCase = new OrderApprovalUseCase(orderRepository);
    });

    test('approve existing order', () => {
        orderRepository.addOrder(new Order(OrderStatus.CREATED, 1));

        useCase.run(new OrderApprovalRequest(1, true));

        const savedOrder = orderRepository.getSavedOrder();
        expect(savedOrder?.status).toBe(OrderStatus.APPROVED);
    });

    test('reject existing order', () => {
        orderRepository.addOrder(new Order(OrderStatus.CREATED, 1));

        useCase.run(new OrderApprovalRequest(1, false));

        const savedOrder = orderRepository.getSavedOrder();
        expect(savedOrder?.status).toBe(OrderStatus.REJECTED);
    });

    test('not approve rejected order', () => {
        orderRepository.addOrder(new Order(OrderStatus.REJECTED, 1));

        expect(() => {useCase.run(new OrderApprovalRequest(1, true))}).toThrowError(RejectedOrderCannotBeApprovedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
    });

    test('not reject approved order', () => {
        orderRepository.addOrder(new Order(OrderStatus.APPROVED, 1));

        expect(() => {useCase.run(new OrderApprovalRequest(1, false))}).toThrowError(ApprovedOrderCannotBeRejectedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
    });

    test('not approve shipped orders', () => {
        orderRepository.addOrder(new Order(OrderStatus.SHIPPED, 1));

        expect(() => {useCase.run(new OrderApprovalRequest(1, true))}).toThrowError(ShippedOrdersCannotBeChangedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
    });

    test('not reject shipped orders', () => {
        orderRepository.addOrder(new Order(OrderStatus.SHIPPED, 1));

        expect(() => {useCase.run(new OrderApprovalRequest(1, false))}).toThrowError(ShippedOrdersCannotBeChangedError);
        expect(orderRepository.getSavedOrder()).toBeUndefined();
    });

});
